Architecture updater (auto-sync) - Updating AArch64

Rot127 opened the pull request:
Draft PR for the auto-sync refactor of AArch64.

- [x] Make Capstone build with generated and translated files.
- [x] Add details
  - [x] Implement operand handling
  - [ ] Fix `td` files
    - [ ] SME vector lead/store/read/write have incorrect outs/ins set
- [x] Update and add tests
- [ ] `za` writes are not listed in register write list
- [x] ~~Fix ARM Disassembler to set Decoder arg~~ Do during LLVM 17 update
- [x] Add `AArch64_GRP_PRIVILEGE` again
- [x] Handle detail in `printSyspAlias` `printSysAlias`
- [x] Handle alias
- [x] Brackets in some names: `ID: 73 (bfmlalb{)`
- [x] Make sure post index works well.
- [x] `'A0 4D DF 4C'` op 4 missing
- [x] `0x21c8df4d` disp missing.
- [x] Missing `#0` imm for `'\x20\x98\xe0\x5e'`
- [x] MIssing imm `3000df4c`

**Todo after merge**

- Fix this 1am decision and rename `HelperMethods.py` to `Helpers.py`.
- New `Decoder` pointer arg must be added to ARM, PPC, TriCore.
- Add `CS_OP_MEM_REG` and `CS_OP_MEM_IMM` op types to prevent confusion that the `CS_OP_MEM` is a flag within the type.
- Generate `CAPSTONE_DIET` builds into all `printOperand()` functions.
- lookupByName for ARM
- [ ] LLVM bug (42e9dec8e927f9d4e9faa160b3d27468e9f5e2ec) for `bfc` alias.
- [ ] U/SBFM alias modulo change depending on S and N bit. (wait for encoding info).

**Breaking changes**

- Renamed all `ARM64` -> `AArch64` (for filenames, enums variable names). Necessary to be consistent with LLVM.
- `SME` operands changed (contin more detail, terminology is closer to the docs).
- System operands change (now categorized into SysAlias, SysImm, SysReg).

**Closes**

closes https://github.com/capstone-engine/capstone/issues/1857
closes https://github.com/capstone-engine/capstone/issues/1853
closes https://github.com/capstone-engine/capstone/issues/1854
closes https://github.com/capstone-engine/capstone/issues/1876 (matches LLVM disassembly now, which is considered source of truth).
closes https://github.com/capstone-engine/capstone/issues/1881
closes https://github.com/capstone-engine/capstone/issues/1887
closes https://github.com/capstone-engine/capstone/issues/1890
closes https://github.com/capstone-engine/capstone/issues/1895
closes https://github.com/capstone-engine/capstone/issues/1953
closes https://github.com/capstone-engine/capstone/issues/2055
closes https://github.com/capstone-engine/capstone/issues/2072
closes https://github.com/capstone-engine/capstone/issues/2130
closes https://github.com/capstone-engine/capstone/issues/1842
closes https://github.com/capstone-engine/capstone/issues/1760
closes https://github.com/capstone-engine/capstone/issues/1670
closes https://github.com/capstone-engine/capstone/issues/1669
closes https://github.com/capstone-engine/capstone/issues/1652
closes https://github.com/capstone-engine/capstone/issues/1629
closes https://github.com/capstone-engine/capstone/issues/1621
closes https://github.com/capstone-engine/capstone/issues/1620
closes https://github.com/capstone-engine/capstone/issues/1591

wait for discussion: https://github.com/capstone-engine/capstone/issues/1889

Rot127 commented:
@kabeor @aquynh @XVilka Any objections against renaming `ARM64` to `AArch64`?

LLVM refers to it as `AArch64` and Capstone is inconsistent with it. So would you agree that we should stick to only one of those names (I'd prefer `AArch64` so we are closer to LLVM)?
Edit: (AArch64 would also mean less `if statements` in the update scripts).

kabeor commented:
@Rot127 Not at all. However, do this mean that users need change their code if they update from v4? (Maybe alias for arm64_xxx like functions needed?)

Rot127 commented:
> However, do this mean that users need change their code if they update from v4?

Yes. Everything which was hand edited in Capstone uses `ARM64` as prefix (file-, enum-, function names etc.). But code from LLVM uses `AArch64`.

If we decide to use `ARM64` or leave the current inconsistent naming, we add complexity to the update scripts and the backends. Because we generate most enums and mapping tables now with the backends, we would need to check the target name every time it is used (and emit `ARM64` instead of `AArch64`).

The decision is basically:

**Change to `ARM64`**

Pro: Saves users to trouble to update their code.
Con: On the cost of adding more (maybe a lot) complexity to `auto-sync`

**Use `AArch64` everywhere**

The other way around.

Edit: Personally I would still like to go the `AArch64` way for easier maintainability. Also there are many changes coming up anyway. So the users are only forced once to adapt their code.
On the other hand, I can not judge how deeply CS is integrated in other projects.

XVilka commented:
No objections against renaming

Rot127 commented:
> Yes. Everything which was hand edited in Capstone uses ARM64 as prefix (file-, enum-, function names etc.). But code from LLVM uses AArch64.

@kabeor @aquynh Any more comments regarding this? Otherwise I would go forward with it.

Rot127 commented:
@kabeor If there is not a really strong argument for capitalized target names in identifiers, I would like to revert this commit (https://github.com/capstone-engine/llvm-capstone/pull/8/commits/cef9d8a6c0a9437560feb4d9870ae950e2a4e1da).
For `AArch64` it gets really difficult to patch most of the usage out of the translated files. Also `AArch64` identifiers weren't capitalized before anyway.

XVilka commented on suite/auto-sync/CppTranslator/Patches/NamespaceArch.py:
```diff
@@ -39,4 +39,12 @@ def get_patch(self, captures: [(Node, str)], src: bytes, **kwargs) -> bytes:
                 case _:
                     res += get_text(src, d.start_byte, d.end_byte) + b"\n"
-        return res
+        return (
```
Will this change also affect previously generated files for Tricore, ARMv7, PPC?

Rot127 replied:
Not semantically. It's only a comment to spot the previous namespace beginning and end. It is meant as a hint that all identifiers within this "scope" should have the namespace id prepended.

In general, all identifiers which were part of a namespace get the namespace id prepended because it eases generation. Since those identifiers are always accessed via the qualified identifier within C++ code we can translate them always like this: `AArch64PacKey::someID` -> `AArch64PacKey_someID`.
This is easier then removing the namespace id at a thousand places.

But yes. I'll update the translated files before the `v6` release so this stuff is in there as well.

Rot127 commented:
@FinnWilkinson While updating `AArcch64` I also do a lot of necessary refactor and renaming while at it. You added the SME extension and I would like to change the Matrix SME op as well. So the naming is more in line with the [SME docs](https://developer.arm.com/documentation/ddi0616/latest).

```c
typedef struct aarch64_op_sme {
  aarch64_sme_op_type type; ///< AArch64_SME_OP_TILE, AArch64_SME_OP_TILE_VEC, AArch64_SME_OP_ACC_MATRIX
  aarch64_reg tile;  ///< Matrix tile register
  aarch64_reg slice_reg; ///< slice index reg
  uint8_t slice_offset;	  ///< slice index offset
} aarch64_op_sme;
```

The `sme_op_types` would follow the [LLVM description](https://github.com/llvm/llvm-project/blob/b0525f66988eac69138ba4ca30f5f07ff5218093/llvm/lib/Target/AArch64/AArch64RegisterInfo.td#L1567-L1595)  of those ops:
- `AArch64_SME_OP_TILE`: Single tile (only `tile` is set).
- `AArch64_SME_OP_TILE_VEC`: Indexed tile (`tile`, `slice_reg`, `slice_offset` is set).
- `AArch64_SME_OP_ACC_MATRIX`: Complete accumulator matrix (only `tile` is set to reg `ZA`)

Could you quickly give a thumbs up if this makes sense?

Rot127 committed:
Update sysop inc file

Rot127 committed:
Fix missing  braces warning

Rot127 committed:
Handle new system operands

Rot127 committed:
Fix build errors by renaming.

Rot127 committed:
Fix segfault

Rot127 committed:
Fix segfault

Rot127 committed:
Add custom MCOperand valiadtors

Rot127 committed:
Add AArch64 case for getFeatureBits

Rot127 committed:
Fix infinite loop

Rot127 committed:
Fix braces warning.

Rot127 committed:
Implement loopuo by name for sys operands

Rot127 committed:
Fix incorrect translation which remove else if statements.

Rot127 committed:
Fix several segfaults

Rot127 committed:
Rename GetRegFromClass patch

Rot127 committed:
Fix segfaults and asserts

Rot127 committed:
Fix segfault

Rot127 committed:
Move MRI setting to Mapping

Rot127 committed:
Remove unused code

Rot127 committed:
Add add_op_X functinos for AArch64.

Rot127 committed:
Add fill detail functins

Rot127 committed:
Handle RegWithShiftExtend operands

Rot127 committed:
Handle TypedVectorList operands.

Rot127 committed:
Handle ComplexRoatation operands

Rot127 committed:
Handle MemExtend operands

Rot127 committed:
Handle ImmRangeScale operands

Rot127 committed:
Handle ExactFPImm operands

Rot127 committed:
Handle GPRSeqPairsClass operands

Rot127 committed:
Handle Imm8OptLsl operands

Rot127 committed:
Handle ImmScale operands

Rot127 committed:
Handle LogicalImm operands

Rot127 committed:
Handle Matrix operands

Rot127 committed:
Handle SME Matrix tiles and vectors.

Rot127 committed:
Handle normal operands.

Rot127 committed:
Fix segfault.

Rot127 committed:
Handle PostInc operands.

Rot127 committed:
Reorder VecLayout enum to have no duplicate enum value.

Rot127 committed:
Handle PredicateAsCounter operands

Rot127 committed:
Handle ZPRasFPR operands

Rot127 committed:
Handle VectorIndex operands

Rot127 committed:
Handle UImm12Offset operands.

Rot127 committed:
Move reg suffix to enum val to single function.

Rot127 committed:
Handle SVERegOp operands

Rot127 committed:
Handle SVELogicalImm operands

Rot127 committed:
Handle SImm operand

Rot127 committed:
Handle PrefetchOp operands

Rot127 committed:
Handle Imm and ImmHex operands

Rot127 committed:
Handle GPR64as32 and GPR64x8 operands

Rot127 committed:
Add missing break

Rot127 committed:
Handle FPImm operand

Rot127 committed:
Handle ExtendedRegister opreand

Rot127 committed:
Handle CondCode operands

Rot127 committed:
Handle BTIHintOp operands

Rot127 committed:
Handle BarrierOption operands

Rot127 committed:
Handle BarrierXSOption

Rot127 committed:
Add not implemeted case again

Rot127 committed:
Handle ArithExtend operands

Rot127 committed:
Handle AdrpLabel and AlignedLabel operands

Rot127 committed:
Handle AMNoIndex operands

Rot127 committed:
Handle AddSubImm operands

Rot127 committed:
Handle MSRSystemRegisters and MRSSystemRegister operands

Rot127 committed:
Handle PSBHntOp and RPRFMOperand operands

Rot127 committed:
Remove unused variables

Rot127 committed:
Handle InverseCondCode operands

Rot127 committed:
Handle ImplicityTypedVectorList operands

Rot127 committed:
Handle ShiftedRegister operands

Rot127 committed:
Handle Shifter operands

Rot127 committed:
Handle SIMDType10Operand operands

Rot127 committed:
Handle SVCROp operands

Rot127 committed:
Handle SVEPattern operands

Rot127 committed:
Handle SVEVecLenSpecifier operands

Rot127 committed:
Handle SysCROperands

Rot127 committed:
Handle SysXzrPair operands

Rot127 committed:
Handle PState operands

Rot127 committed:
Handle VRegOperands

Rot127 committed:
Primt SME oeprands.

peace-maker committed:
Add Python bindings for SH

XVilka commented:
Please rebase on top of the latest `next`

Rot127 commented:
Do it after PPC is in a more polished state (in a few days).

peace-maker committed:
Fix ARM Python bindings (#2127)

Rot127 committed:
Fix cs_operand.h include

Rot127 committed:
Rename arm64 -> aarch64 in python bindings.

Rot127 committed:
Restructure auto-sync update scripts.

Rot127 committed:
Move Helper functions to Updater dir

Rot127 committed:
Move requirements.txt

Rot127 committed:
Add basic ASUpdater.py

Rot127 committed:
Run black.

Rot127 committed:
Add inc file generater to updater

Rot127 committed:
Add option to select certain inc files fore generation.

Rot127 committed:
Enable clean build and implement patcher for inc files.

Rot127 committed:
Format config

Rot127 committed:
Patch main header files after inc generation.

Rot127 committed:
Implement clang-format function (unused yet, because it takes forever.)

Rot127 committed:
Copy generated inc files to arch dir

Rot127 committed:
Invert clean option (noramlly we need to clean the build dir.)

Rot127 committed:
Clearify arg doc

Rot127 committed:
Rename SystemRegister file for AArch64

Rot127 committed:
Centralize handling of path variables.

Rot127 committed:
Check if SystemOperands had to be generated before renaming on of its files.

Rot127 committed:
Replace class parameters by calling get_path

Rot127 committed:
Remove updater config which only contained paths.

Rot127 committed:
Add refactor option.

Rot127 committed:
Remove more path handling in the Configurator.

Rot127 committed:
Add translation step to updater.

Rot127 committed:
Fix includes after CppTranslator was moved into the Updater

Rot127 committed:
Remove updater config

Rot127 committed:
Fix several issue in the Configurator

Rot127 committed:
Fix file operations

Rot127 committed:
Remove addition argument from translator.

Rot127 committed:
Add Differ step to updater.

Rot127 committed:
Add path variable for arch_config

Rot127 committed:
Add diff step.

Rot127 committed:
Fix typo

Rot127 committed:
Introduce .clang-format path variable.

Rot127 committed:
Remove duplicate functions

Rot127 committed:
Add option to select update steps to execute.

Rot127 committed:
Check in write functions for write flag.

Rot127 committed:
Rename PatchMainHeader -> HeaderPatcher

Rot127 committed:
Move .gitignore

Rot127 committed:
Add README to vendor dir.

Rot127 committed:
Add all system operands to cstool output

Rot127 committed:
Update cstest with aarch64 changes

Rot127 committed:
Remove wb flag of aarch64 detail struct

Rot127 committed:
Set updates_flag after decoding

Rot127 committed:
Set writeback after decoding.

Rot127 committed:
Rename ARM64 -> AArch64

Rot127 committed:
Update printer and op mapping

Rot127 committed:
Exit normally

Rot127 committed:
Add AArch64 alias

Rot127 committed:
Fix some tmeplate function calls

Rot127 committed:
Fix flag check after rebase.

Rot127 committed:
Fix build by commentig unnused code.

Rot127 committed:
Add memory operand flag

Rot127 committed:
Handle memory operands printed via generic printOperand function.

Rot127 committed:
Handle UImm memory offsets

Rot127 committed:
Introduce MEM_REG and MEM_IMM op types

Rot127 committed:
Handle scaled memory immediates

Rot127 committed:
Check for op_count before checking for mem op at -1 index.

Rot127 committed:
Update memory operand flags.

Rot127 committed:
Pass imm/reg memory ops in set_imm/reg to set_mem.

Rot127 committed:
Add missing set_sme_operand call and fix assert.

Rot127 committed:
Remove CS_OP_MEM flag before entering switch.

Rot127 committed:
Preidcates are registers.

Rot127 committed:
Add shift info always to the previous operand

Rot127 committed:
Check for generic system regs

Rot127 committed:
Handle NumLanes = 0 LaneKind = q case

Rot127 committed:
Replace printImm call with normal print logic. Otherwise ops get added twice to detail.

Rot127 committed:
Handle FP operands in printOperand.

Rot127 committed:
Add access information to float operands.

Rot127 committed:
Rewrite SME matrix handling.

Rot127 committed:
Set correct SME layouts and allow for immediate range sme offsets.

Rot127 committed:
Handle cases of unknown system alias by setting their raw values

Rot127 committed:
Update cstool and header file with new SME offset handling

Rot127 committed:
Handle SME Tile lists.

Rot127 committed:
Fix build error in cstest

Rot127 committed:
Update MC tests for AArch64

Rot127 committed:
Handle TLBI operands and fix printing bug.

Rot127 committed:
Fix: Print signed value as signed.

Rot127 committed:
Add more system alias to detail.

Rot127 committed:
Remove duplicate hex prefix

Rot127 committed:
Set correct values for the register info

Rot127 committed:
Replace tabs with white spaces

Rot127 committed:
Move string append logic to own function.

Rot127 committed:
Set DecodeComplete = true before decoding (as originally in the LLVM code).

Rot127 committed:
Change type of feature argument, since only LLVM features are passed, not CS groups.

Rot127 committed:
Imitate lower_bound for the index table binary search.

Rot127 committed:
Remove trailing comments from test files.

Rot127 committed:
Print shift amount in decimal

Rot127 committed:
Save detail of shift alias instructions.

Rot127 committed:
Add extension details fot ext instruction alias

Rot127 committed:
Print LSB and width in decimal

Rot127 committed:
Fix LLVM bug. The feature check for V8_2a doesn't check if all features are enabled.

Rot127 committed:
Fix lower_bounds check.
For m == 0 we wrap around 0 of cause.

Rot127 committed:
Fix feature check. Add check for FeatureAll since it includes XS

Rot127 committed:
Operate on temporary MCInst when trying decoding.

Rot127 committed:
Add lower_bound behavior to IndexTypeStr binsearch.

Rot127 committed:
Fix MC tests which were incorrect because of missing FeatureAll check

Rot127 commented:
@kabeor @aquynh The disassembler tests run through so far and every possible instruction was decoded at least once without segfault. Going to continue with Rizin integration now, which will test the details set (CI still fails because of the yet unfixed `issues.cs`). But in general everything should be in a good shape.
Feel free to start your review.

Rot127 committed:
Add Alias handling for AArch64

Rot127 committed:
Update system operands with SYSIMM types and add additional sysop category.

Rot127 committed:
Add macros for meta programming (ARM64 <-> AArch64 selection).

Rot127 committed:
Fix union/struct confusion and add raw_value member to uninions.

Rot127 committed:
Allow to set Syntax and mode options for AArch64

Rot127 committed:
Print shift value in decimal

Rot127 committed:
Fix build warning by using correct type

Rot127 committed:
Add missing call to add_cs_detail.

Rot127 committed:
Update name map files with normalized names.

Rot127 committed:
Remove unused function

Rot127 committed:
Add check if detail should be filled.

Rot127 committed:
Fill detail for real instructions if only real detail is requested.

Rot127 committed:
Add always the extension.

watbulb committed:
add AARCH64 to `cmake.sh`

watbulb committed:
add HAS_AARCH64 to `cs.c`

watbulb committed:
should probably just reference `cs_operand.h` in `aarch64.h`

watbulb committed:
hint compiler at `AArch64_SYSREG` enum type for casting purposes

watbulb committed:
update `Makefile` for AARCH64

leaves `CAPSTONE_HAS_ARM64` supported

watbulb committed:
`testFeatureBits` platform function check

`testFeatureBits` should check if the platform function is visible first

watbulb committed:
update tests to use AARCH64 convention

watbulb committed:
hack: avoid enum casts for `MCInst` Values

Apple compiler really hates typecasting a enum, even if bounded from a unsigned. Lets set the raw_value directly

is a hack and needs proper review

Rot127 committed:
Make dir creation log message debug level

Rot127 committed:
Implement ADR immediate operand printer.

See: https://github.com/capstone-engine/llvm-capstone/commit/c3484b1fdc03b479beaf5897eca8ea294d3df909

Rot127 committed:
Check for flag registers beeing written and update flag.

Rot127 committed:
Move multiple CondCode helpers to aarch64.h because they are so freaking useful.

+ Print CC if it is EQ

Rot127 committed:
Fix incorrectly initialized CC and VectorLayout.

Rot127 committed:
Add LSL shift type for extensions.

Rot127 committed:
Fix case when shift amount is 0

Rot127 committed:
Fix post-index memory instructions.

Rot127 committed:
Pass raw immediate through getShiftValue to extract actual shift amount

Rot127 committed:
Setup AArch64 detail ops.

Rot127 committed:
Add flag for operands part of a list.

Rot127 committed:
Set vector indices for all relevant registers.

Rot127 committed:
Add missing call to add_cs_detail for postIncOperands

Rot127 committed:
Add ugly yet reliable way to determine post-index addressing mode

Rot127 committed:
Add support for old Capstone register alias.

Rot127 committed:
Remove leading space before some alias mnemonics.

XVilka commented:
Since reviews take so long, I recommend starting them right now for this PR @kabeor @aquynh

FinnWilkinson commented:
> @FinnWilkinson While updating `AArcch64` I also do a lot of necessary refactor and renaming while at it. You added the SME extension and I would like to change the Matrix SME op as well. So the naming is more in line with the [SME docs](https://developer.arm.com/documentation/ddi0616/latest).
> 
> ```c
> typedef struct aarch64_op_sme {
>   aarch64_sme_op_type type; ///< AArch64_SME_OP_TILE, AArch64_SME_OP_TILE_VEC, AArch64_SME_OP_ACC_MATRIX
>   aarch64_reg tile;  ///< Matrix tile register
>   aarch64_reg slice_reg; ///< slice index reg
>   uint8_t slice_offset;	  ///< slice index offset
> } aarch64_op_sme;
> ```
> 
> The `sme_op_types` would follow the [LLVM description](https://github.com/llvm/llvm-project/blob/b0525f66988eac69138ba4ca30f5f07ff5218093/llvm/lib/Target/AArch64/AArch64RegisterInfo.td#L1567-L1595) of those ops:
> 
> * `AArch64_SME_OP_TILE`: Single tile (only `tile` is set).
> * `AArch64_SME_OP_TILE_VEC`: Indexed tile (`tile`, `slice_reg`, `slice_offset` is set).
> * `AArch64_SME_OP_ACC_MATRIX`: Complete accumulator matrix (only `tile` is set to reg `ZA`)
> 
> Could you quickly give a thumbs up if this makes sense?

> 

@Rot127  Appologies that I haven't seen this sooner... I've been watching this PR for a while and have completely missed this. 

This looks good yes for SME. I presume that for SME2's new multi-vector operations & vector-groups this is encoded into the instruction enum rather than neededing to be another field of the `aarch64_op_sme` struct? 
Example usage can be seen in this [SUDOT](https://developer.arm.com/documentation/ddi0602/2023-06/SME-Instructions/SUDOT--multiple-and-indexed-vector---Multi-vector-signed-by-unsigned-integer-dot-product-by-indexed-elements-) SME2 instruction.

Rot127 commented:
@FinnWilkinson Currently we can not distinguish between `vgx2` and `vgx4` sadly. Capstone's instr. enums match always the mnemonic. The LLVM instr. enums do differ between, but there are just too many (easily 4000+) to be practical usable.

But we can add the property easily by checking the LLVM instr. id in Capstone and set a flag somewhere.

The vector groups seem more a property of the matrix operand. Hence, I would prefer adding a new field to the `sme` operand. You seem one of the few who has a need for this extensions. So would you suggest something else (and for what reason)?

FinnWilkinson commented:
@Rot127 I see, sorry I confused the Capsone and LLVM enums there (as we use the LLVM ones for our tool). 

And yes, I think adding a field to the `sme` operand to indicate vector group is a good idea, where `vector_groups` can have a value of `1`, `2`, or `4` to indicate absece of `VG` (aka single vector-group), presence of `VGx2`, or  presence of `VGx4` respectively (or something similar). 

Adding this to cstool's `print_insn_detail_aarch64` would be good, as well as adpating `AArch64InstrPrinter.c` to add the `VGx2`/`VGx4` to the `op_str` to make it clear in disassembly exactly which variation of the instruction is present. A specific test case to make sure everything is printing as it should could be [FMLSL](https://developer.arm.com/documentation/ddi0602/2023-06/SME-Instructions/FMLSL--multiple-vectors---Multi-vector-floating-point-multiply-subtract-long-) given it is widening, has a range offset, and has vector-groups.

XVilka commented:
Because for some reason I cannot comment in https://github.com/Rot127/capstone/pull/1, adding it here:

@watbulb @Rot127 In general, it's better to use `{}` braces even with one-line conditions - it will help avoiding indentation-related mistakes in the future. Same in other places.

Rot127 committed:
Check for present detail before accessing it.

Rot127 committed:
Add CS only groups

Rot127 committed:
Use general map ins_op type

Rot127 committed:
Fix build warning about str size computation.

Rot127 committed:
Disable warning about unitialized value for GCC 11.

Imm is initialized and the warning does not appear
in later versions.

Rot127 committed:
Use correct include guard for PPC

Rot127 committed:
Add missing requirements

Rot127 committed:
Update SystemOperand enums.

Rot127 committed:
Fix overlapping comparison warning

Rot127 committed:
Fix reachable assert where OpNum is not of type IMM

Rot127 committed:
Handle 0.0 operand for fcmp

Rot127 committed:
Fix incorrect variable passed.

Rot127 committed:
Fix for MacOS which doesn't know the warning and throws another one.

Rot127 committed:
Make getExtendEncoding static to fix build warning on MSVC.

Rot127 committed:
Fix build error: 'missing binary operator before token' by checking __GNUC__

Rot127 committed:
Add string search to add vector layout info.

Rot127 committed:
Add missing mem disponents of several ldr and str instructions.

Rot127 committed:
Add 0 immediates to several instructions.

Rot127 committed:
Rename v regs to q and d variant.

The cs_regname API can not pass the variant name of the register requested.
So we simply emit the default variant name.

Rot127 committed:
Fix incorrect enum value.

Rot127 committed:
Fix tests for system operands.

Rot127 committed:
Fix syntax issues in tests.

Rot127 committed:
Rename Arm64 -> AArch64 Python bindings.

Rot127 committed:
Fix Python bindings C structs.

Rot127 committed:
Fix generation of constants (ARMCC skipped because it starts with ARM)

Rot127 committed:
Update const files

Rot127 committed:
Remove -Wmaybe-uninitialized warning since it fails fuzz build

Rot127 committed:
Add missing comma

Rot127 committed:
Fix case

Rot127 committed:
Fix AArch64 Python bindings:

- Do not generate constants automatically (dscript is way too buggy).
- Update printing of details.

Rot127 committed:
Rename ARM64 -> AArch64 in test_corpus.py

Rot127 committed:
Rename test_arm64 -> test_aarch64

Rot127 committed:
Rename ARM-64 -> AArch64

Rot127 commented:
@kabeor @aquynh  I open this now for review so you have enough time.

There are still minor things left. Mostly syntax changes in the Python tests (hence the failing tests) and some naming problems (the failing fuzz test). I will run the complete instruction space once through it in the next days to check for segfaults again. But generally the `issues.cs` pass and this will not change much.

I'll add the encoding after this is merged. Since this is already huge.

Rot127 committed:
Fix diff CI test by disassembling AArch64 at former ARM64 place

Rot127 committed:
Fix several wrong types and remove unnecessary memebers from Python binding

Rot127 committed:
Fix: Same printing format of detail for cstool, test_ and test_*.py

Rot127 committed:
Fix: pass correct op index for mov alias with op[1] == reg wzr.

Rot127 committed:
Set prfm op manuall in case of unnown sysop. set_imm would add it to an memory operand wihtout base.

Rot127 committed:
Fix: If barrier ops are not set an assert is reached.

We fix it here by simply getting the immediate as the printing code does.

Rot127 commented:
`0x0` - `0xffffffff` go all through now (with `-d` and without `-d`).

Rot127 commented:
@kabeor @aquynh I think I'd like to address the last two points on the list with LLVM 17. Since it needs an update anyways. Besides that, I would really like to see this merged. Then we got the biggest part of this whole thing behind us.

kabeor commented:
Will merge it this weekend:)

kabeor commented:
@Rot127 LGTM, can you make a warning to ask them to upgrade their code when people who are still calling arm64xxx? Will merge after that.

Rot127 commented:
@kabeor Where would you add such a warning?
Also mind that there is still the [V6 guide](https://github.com/capstone-engine/capstone/pull/2124).
What do you think about finishing the guide and make a v6 alpha release?
In the release page all the big changes could be listed as well as missing features (encoding, real/alias detail ARM etc.). This would also bring to attention, that there is quite some change going on.

XVilka commented:
I think such a warning, if necessary, should be done in a separate PR

XVilka commented:
@kabeor @aquynh, what do you think about merging this as is and further improvements in the following PRs? This one grew very big already, and merging it will help more widespread testing of this code.

kabeor commented:
Well I mean a warning comes from code. 

For example, if a user directly upgrades to v6 without reading the documentation, when he still calls the previous code (containing arm64 API), the code should return a warning and stop running.

It's ok to be done in a separate PR, can I assign this to you? @Rot127 

And please confirm, if there is nothing still being done in this PR, I will merge now.

Rot127 commented:
> And please confirm, if there is nothing still being done in this PR, I will merge now.

You can merge.

> For example, if a user directly upgrades to v6 without reading the documentation, when he still calls the previous code (containing arm64 API), the code should return a warning and stop running.

Since we changed all the names, people wouldn't be able to build Capstone. Or are you speaking of prebuilds?

In the latter case yes sure. For `cstool` we could add a warning if someone passes `arm64`. You can assign this to me. `libcapstone` again will fail to link, since everything is renamed now.

But this is also why I think we should make some kind of pre-release. So everyone is aware and knows the details.

kabeor commented:
Cool, merged!
